src/core/main.c

```diff
@@ -2004,15 +2004,24 @@ static void log_execution_mode(bool *ret_first_boot) {
                         *ret_first_boot = false;
                         log_info("Running in initial RAM disk.");
                 } else {
-                        /* Let's check whether we are in first boot, i.e. whether /etc is still unpopulated. We use
-                         * /etc/machine-id as flag file, for this: if it exists we assume /etc is populated, if it
-                         * doesn't it's unpopulated. This allows container managers and installers to provision a
-                         * couple of files already. If the container manager wants to provision the machine ID itself
-                         * it should pass $container_uuid to PID 1. */
-
-                        *ret_first_boot = access("/etc/machine-id", F_OK) < 0;
-                        if (*ret_first_boot)
-                                log_info("Running with unpopulated /etc.");
+                        int r;
+                        _cleanup_free_ char *id_text = NULL;
+
+                        /* Let's check whether we are in first boot.  We use /etc/machine-id as flag file
+                         * for this: If it is missing or contains the value "uninitialized", this is the
+                         * first boot.  In any other case, it is not.  This allows container managers and
+                         * installers to provision a couple of files already.  If the container manager
+                         * wants to provision the machine ID itself it should pass $container_uuid to PID 1. */
+
+                        r = read_one_line_file("/etc/machine-id", &id_text);
+                        if (r < 0) {
+                                *ret_first_boot = true;
+                                log_info("Running for the first time.");
+                        } else if (strcmp(id_text, "uninitialized") == 0) {
+                                *ret_first_boot = true;
+                                log_info("Running for the first time (detected previous incomplete boot).");
```

Comment:
> I totally see usecases where one would initialize the file like this on a deployment image.

I also had this thought, but actually, I'm not sure where this would make sense.  `uninitialized` behaves as if the file was missing entirely, so any image which wants to have the first-boot detection should deploy without `/etc/machine-id`.

For a read-only rootfs it also doesn't make much sense, as this would mean systemd always thinks it's booting for the first time.

But I can remove the message, I agree that it might be better without it.

Reply:
but it would allow people to start with a read-only root, that is then remounted later on writable, and consider that a first boot scenario. That is very useful. After all systemd-machine-id-commit has this beautiful logic of commiting the machine ID atomically after the root disk has been mounted writable